src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php

```diff
@@ -252,6 +254,54 @@ private function validateAndDenormalize($currentClass, $attribute, $data, $forma
                 $data = [$data];
             }
 
+            // In XML and CSV all basic datatypes are represented as strings, it is e.g. not possible to determine,
+            // if a value is meant to be a string, float, int or a boolean value from the serialized representation.
+            // That's why we have to transform the values, if one of these non-string basic datatypes is epxected.
+            //
+            // This is special to xml and csv format
+            if (
+                \is_string($data) &&
+                (XmlEncoder::FORMAT === $format || CsvEncoder::FORMAT === $format) &&
+                \in_array($type->getBuiltinType(), [Type::BUILTIN_TYPE_BOOL, Type::BUILTIN_TYPE_INT, Type::BUILTIN_TYPE_FLOAT], true)
+            ) {
+                switch ($type->getBuiltinType()) {
+                    case Type::BUILTIN_TYPE_BOOL:
+                        // according to https://www.w3.org/TR/xmlschema-2/#boolean, valid representations are "false", "true", "0" and "1"
+                        if ('false' === $data || '0' === $data) {
+                            $data = false;
+                        } elseif ('true' === $data || '1' === $data) {
+                            $data = true;
+                        }
+                        break;
+                    case Type::BUILTIN_TYPE_INT:
+                        if (
+                            ctype_digit($data) ||
+                            '-' == $data[0] && ctype_digit(substr($data, 1))
+                        ) {
+                            $data = (int) $data;
+                        }
+                        break;
+                    case Type::BUILTIN_TYPE_FLOAT:
+                        if (is_numeric($data)) {
+                            return '0x' === $data[0].$data[1] ? hexdec($data) : (float) $data;
```

Comment:
The `0x` case is not possible:

> Hexadecimal (e.g. 0xf4c3b00c) and binary (e.g. 0b10100111001) notation is not allowed.

https://www.php.net/manual/en/function.is-numeric.php

Reply:
Also, `xsd:float` does not allow using hexadecimal representation anyway:

https://www.w3.org/TR/xmlschema-2/#float